### PR TITLE
feat: implement the Bohr upgrade of BSC

### PR DESCRIPTION
**Pull request by yutianwu**
### Description

This pr will implement the below beps of BSC and introduce the `Bohr` upgrade:

https://github.com/bnb-chain/BEPs/pull/341
https://github.com/bnb-chain/BEPs/pull/402
https://github.com/bnb-chain/BEPs/pull/404

You can refer to the PRs of BSC:

https://github.com/bnb-chain/bsc/pull/2482
https://github.com/bnb-chain/bsc/pull/2502
https://github.com/bnb-chain/bsc/pull/2558

### Rationale

Implement the BEPs of BSC

### Example

n/a

### Changes

Notable changes: 
* implement the Bohr upgrade of BSC

### Potential Impacts

n/a

### CODE REVIEW

File: crates/primitives/src/parlia/snapshot.rs [hunk 1]
```diff
@@ -8,6 +8,8 @@ use std::collections::{BTreeMap, HashMap};
 
 /// Number of blocks after which to save the snapshot to the database
 pub const CHECKPOINT_INTERVAL: u64 = 1024;
+/// Default turn length before Bohr upgrade
+pub const DEFAULT_TURN_LENGTH: u8 = 1;
```
Comment (pythonberg1997) [inline crates/primitives/src/parlia/snapshot.rs:12]: There're 2 `DEFAULT_TURN_LENGTH` definition

Comment (yutianwu) [reply crates/primitives/src/parlia/snapshot.rs:12]: They are not the same, one is the size of the variable, one is the default value of turn_length

File: crates/bsc/consensus/src/lib.rs [hunk 2]
```diff
@@ -289,10 +315,15 @@ impl Parlia {
             }
 
             // Exclude the recently signed validators
-            validators.retain(|addr| !recents.contains_key(addr));
+            validators.retain(|addr| !snap.sign_recently_by_counts(*addr, &counts));
```
Comment (unclezoro) [inline crates/bsc/consensus/src/lib.rs]: on geth, there are code like:
```
				if p.chainConfig.IsBohr(header.Number, header.Time) {
					if addr == inTurnAddr {
						continue
					}
				}
```

I did not see it in Reth

File: crates/bsc/evm/src/execute.rs [hunk 3]
```diff
@@ -867,21 +867,37 @@ where
...
         // apply skip headers
         skip_headers.reverse();
         for header in &skip_headers {
-            let ValidatorsInfo { consensus_addrs, vote_addrs } = if header.number > 0 &&
-                header.number % self.parlia.epoch() == (snap.validators.len() / 2) as u64
+            let (ValidatorsInfo { consensus_addrs, vote_addrs }, turn_length) = if header.number > 0 &&
+                header.number % self.parlia.epoch() == snap.miner_history_check_len()
             {
                 // change validator set
                 let checkpoint_header =
```
Comment (unclezoro) [inline crates/bsc/evm/src/execute.rs:883]: on geth they get code, dont we need it?
```
			if chainConfig.IsBohr(header.Number, header.Time) {
				// after switching the validator set, snap.Validators may become larger,
				// then the unexpected second switch will happen, just skip it.
				if _, ok := snap.Recents[epochKey]; ok {
					continue
				}
			}
```

File: crates/bsc/consensus/src/lib.rs (hunk 2 above)
Comment (yutianwu) [reply crates/bsc/consensus/src/lib.rs]: you are right

File: crates/bsc/consensus/src/abi.rs [hunk 4]
```diff
@@ -4494,6 +4507,19 @@ lazy_static! {
             ],
             "stateMutability": "view"
           },
+          {
```
Comment (NathanBSC) [inline crates/bsc/consensus/src/abi.rs]: no getTurnLength for STAKE_HUB_ABI

**Review by NathanBSC — APPROVED** [review]

**Review by buddh0 — APPROVED** [review]

File: crates/bsc/evm/src/execute.rs (hunk 3 above)
Comment (buddh0) [reply crates/bsc/evm/src/execute.rs:883]: > on geth they get code, dont we need it?
> 
> ```
> 			if chainConfig.IsBohr(header.Number, header.Time) {
> 				// after switching the validator set, snap.Validators may become larger,
> 				// then the unexpected second switch will happen, just skip it.
> 				if _, ok := snap.Recents[epochKey]; ok {
> 					continue
> 				}
> 			}
> ```

here
image

**Review by unclezoro — APPROVED** [review]

**Review by monkeydluffyto — APPROVED** [review]

**Pull request merged**